Enhance the checks before accessing the memory for the heap heuristics

lebr0nli opened the pull request:
We used `pwndbg.gdblib.vmmap.find()` to verify whether some addresses are valid or not in the heap heuristics.
But in some edge cases, it's still possible that an address can be found by vmmap, but still cause an error when we tried to access it.

This can be reproduced by something like this:
```text
pwndbg> ipi

In [1]: pwndbg.gdblib.vmmap.find(0xffffffffff600000)
Out[1]: Page('0xffffffffff600000 0xffffffffff601000 r-xp     1000      0 [vsyscall]')

In [2]: pwndbg.gdblib.memory.byte(0xffffffffff600000)
---------------------------------------------------------------------------
MemoryError                               Traceback (most recent call last)
<ipython-input-2-410a4ed1fe80> in <module>
----> 1 pwndbg.gdblib.memory.byte(0xffffffffff600000)

/ctf/work/pwndbg/pwndbg/gdblib/memory.py in byte(addr)
    162     Read one byte at the specified address
    163     """
--> 164     return readtype(pwndbg.gdblib.typeinfo.uchar, addr)
    165
    166

/ctf/work/pwndbg/pwndbg/gdblib/memory.py in readtype(gdb_type, addr)
     77         :class:`int`
     78     """
---> 79     return int(gdb.Value(addr).cast(gdb_type.pointer()).dereference())
     80
     81

MemoryError: Cannot access memory at address 0xffffffffff600000
```

This PR aims to prevent it and fixes #1542.

lebr0nli committed:
Enhance the checks before accessing the memory

- Use `pwndbg.gdblib.memory.peek()` instead of `pwndbg.gdblib.vmmap.find()` to check if the address is valid
- Directly access the memory when searching the `main_arena` in memory and catch the exception

lebr0nli committed:
Make finding `main_arena` in memory more efficient and reliable

We only try the address that is aligned to `pwndbg.gdblib.arch.ptrsize`

disconnect3d commented:
@lebr0nli heh, perhaps we should have a version of vmmap that would not return the vsyscall page to be used for cases like here... :)

A better example for this change is when we just can't figure out the vmmap result and return a page of [0, 0xffff...] though the question is --- if that is the case --- do we ever reach those code paths here?

Also, using `memory.peek` instead of `vmmap.get` can be problematic for embedded targets when accessing the memory _is not cheap and is very slow_. However, I think that those targets will likely not be used with libc heap very often, so its probably okay to not think about this until someone actually hits that case.

disconnect3d commented on pwndbg/heap/ptmalloc.py:
```diff
@@ -1408,18 +1419,18 @@ def main_arena(self):
                                 tmp_next = int(tmp_arena["next"])
-                            else:
+                            except Exception:
                                 # if `&tmp_arena->next` is not valid, the linked list is broken, break this while loop and try `addr+pwndbg.gdblib.arch.ptrsize` again
```
Just to be sure: this comment here is still valid after we switch to except, right?

lebr0nli replied:
Yes, it should still be valid.

disconnect3d commented on pwndbg/heap/ptmalloc.py:
```diff
@@ -1408,18 +1419,18 @@ def main_arena(self):
+                            try:
                                 tmp_next = int(tmp_arena["next"])
-                            else:
+                            except Exception:
```
is this gdb.MemoryError or gdb.error that happens here? Can we be more explicit with the exceptions that we want to catch? 🤐

lebr0nli replied:
I think it should be only `gdb.MemoryError`, but I’m not very sure actually.
I was following the code here when writing it:
https://github.com/pwndbg/pwndbg/blob/afa96abfa022052adbf39a430a7f353faac32838/pwndbg/gdblib/memory.py#L110-L114
Is there any reason why we catch `Exception` here?
If it also should be only `gdb.MemoryError`, maybe I should change it as well? (probably change it in another PR?)

disconnect3d commented on pwndbg/heap/ptmalloc.py:
```diff
@@ -1398,7 +1402,14 @@ def main_arena(self):
-                        tmp_next = int(tmp_arena["next"])
+                        try:
+                            tmp_next = int(tmp_arena["next"])
+                        except Exception:
```
is this gdb.MemoryError or gdb.error that happens here? Can we be more explicit with the exceptions that we want to catch? 🤐

disconnect3d commented:
I know that I ask about a lot: but can we maybe add a testcase for this? I bet this can be tested by running some heap command against a vsyscall address page?

lebr0nli commented:
> A better example for this change is when we just can't figure out the vmmap result and return a page of [0, 0xffff...] though the question is --- if that is the case --- do we ever reach those code paths here?

Yes, for the qemu-user without the patch, I guess we might face a similar problem if with only do the vmmap check.

> Also, using `memory.peek` instead of `vmmap.get` can be problematic for embedded targets when accessing the memory _is not cheap and is very slow_

Ahh, I forgot about the performance issues, thanks for pointing it out.
Hmm, maybe for now, using `vmmap.find()` before `memory.peek()` will be better?
So we can avoid some unnecessary memory access if possible.
I mean something like this:
```python
def is_peekable_address(address):
    return pwndbg.gdblib.vmmap.find(address) is None and peek(address) is not None
```

> I know that I ask about a lot: but can we maybe add a testcase for this?

Sure, I'll the test for it.

disconnect3d commented:
> Yes, for the qemu-user without the patch, (...)

I really need to get back to this patch and finish it :).

> Hmm, maybe for now, using vmmap.find() before memory.peek() will be better?

This makes sense! Shouldn't it actually be:
```py
def is_readable_address(address):
    return pwndbg.gdblib.vmmap.find(address) and peek(address)
```
(or, if it makes more sense: `is not None` in both)

I mean, so that we will peek only if we find memory page. This would at least be faster if someone defines the vmmap pages properly on embedded targets.

lebr0nli commented:
> (or, if it makes more sense: `is not None` in both)

Yes, that's right, somehow I left out an `not`, what did I write lol😂

lebr0nli committed:
Avoid unnecessary memory accessing if possible

- Before we used `pwndbg.gdblib.memory.peek()` to check if an address is readable for GDB, we used `pwndbg.gdblib.vmmap.find()` to make sure that this address is in one of the pages, since accessing memory for embedded targets might be slow and expensive
- Create a new function: `is_readable_address` for `pwndbg.gdblib.memory`

lebr0nli committed:
Fix wrong test for `main_arena`

The heap object should be reset before testing the multi-threaded condition

lebr0nli committed:
Add the test to make sure the heap heuristics won't be affected by the vmmap result

Previously, we used `pwndbg.gdblib.vmmap.find()` to check whether the address is valid or not, but this might be a false positive for the address in the `[vsyscall]` page or in the page with a range from 0~0xffffffffffffffff (e.g. qemu-user).
This commit aims to include this scenario during the tests, to make sure the heap heuristics won't be affected by this.

lebr0nli committed:
Use `gdb.MemoryError` instead of `Exception`

disconnect3d commented:
@lebr0nli Is that work ready? I haven't tested the tests myself but I believe they work fine since CI builds! :P

lebr0nli commented:
> Is that work ready?

Yea, I think it's ready to be merged.
